### PR TITLE
Added query.add_rel_alias()

### PR DESCRIPTION
**Pull request by akaariai**
The idea is that one can control explicitly join generation with
add_rel_alias. For example:

    qs.add_rel_alias(friends1='friends', friends2='friends').filter(
       friends1__age__lte=50, friends2__height__gte=180)

is the same query as

    qs.filter(friends__age__lte=50).filter(friends__height__gte=180)

Notably there are certain queries that can't be written without
add_rel_alias(). For example

    qs.add_rel_alias(friends1='friends', friends2='friends').filter(
        ~Q(friends1__age__lte=50, friends2__height__gte=180))

Unforuntaely there are queries we can't handle yet. Any query that
would require pushing two multijoins into a single subquery fail
loudly currently. For example the query

    qs.exclude(Q(friends1__age=50) | Q(friends2__height=180),
               Q(friends1__height=180))

will fail.

The add_rel_alias() can in future be used for limiting join results,
for example

    qs.add_rel_alias(
        translation_fi=Relation('translations', only=Q(lang='fi')))

would effectively add the join

    JOIN translations ON translations.article_id = article.id AND
        translations.lang = 'fi'

to the query.

### CODE REVIEW

**Comment by akaariai** [comment]
This is work on top of https://github.com/django/django/pull/4385.

File: docs/releases/1.10.txt [hunk 1]
```diff
@@ -375,6 +375,31 @@ Dropped support for PostgreSQL 9.1
...
+    Author.objects.exclude(friends__age__gte=20, friends__age__lte=40)
+
+In Django 1.10, any author with a friend age 20 to 40 will be removed from the
+results. In previous versions, any author with a friend over age 20 and a
+(**possibly different**) friend under age 40 are removed from the results.
+
+If you want a query that matches the old behavior of the above query and works
+equivalently on both older and newer version of Django, try::
+
+    Author.objects.exclude(friends__age__gte=20).filter(friends__age__lte=40)
```
Comment (charettes) [inline docs/releases/1.10.txt]: `exclude(friends__age__lte=40)`?

**Comment by akaariai** [comment]
I wonder if we could get away with something a little less drastic than the deferred filtering approach in this PR.

The original problem is that .exclude(friends__age__gte=30, friends__height__lte=180) currently targets two different sets of friends, but according to our API both of the conditions should target the same set of friends. PR https://github.com/django/django/pull/4385 fixes this, but then we don't have any way of writing the original query. If you use two excludes chained, you can't get the two conditions connected by NOT.

Instead of adding the add_rel_alias, we could have a way to tell Django that these filters inside the same query should target different sets of joins. Maybe some marker class for filter/exclude, for example .exclude(models.NewAlias(friends__age__gte=30), friends__height__lte=180), or just .exclude(friends_1__age__gte=30, friends_2__height__lte=180).

Yet another possibility is to do .add_relation(friends1='friends') adds a join for friends to the query, and if you try to exclude against it you get a failure. .add_relation(friends1=models.SubQuery('friends')) adds a subquery for friends1.

The reason I'm not too happy about the current patch is that deferred filters seem to be really hard to combine with annotations. Because we don't know if a relation is going to be subqueried or not until we have seen all the filters/excludes targeting that relation, we can't add annotations against that relation either. In addition, if you have multiple multivalued joins in a single query aggregations aren't going to work.

I'm going to let this rest a little. Maybe some ideas surface.

**Comment by akaariai** [comment]
A random collection of problems with the way we handle multivalued relations: https://gist.github.com/akaariai/716dcdd895ba159123c4

**Comment by akaariai** [comment]
Reminder for myself:
  1. The correct way to query .exclude(Q(friends__age__gte=50), ~Q(friends__height__gte=180)) seems to be to generate a NOT IN subquery, and invert the inner query.
  2. Do not try to support combinations of two different multivalued relations mixed in a single filter clause. Such queries do not make much sense anyways.

**Comment by lucaswiman** [comment]
@akaariai: One issue [I ran into recently](https://groups.google.com/forum/#!topic/django-users/PP1XuhmWcxs) is that even simple queries don't obey De Morgan's law. The following assertions are true on Django <= 1.9, and also on this branch:
```python
test_obj = TestObj.objects.create()
test_obj.m2ms.create(int_value=10, char_value='foo')
test_obj.m2ms.create(int_value=20, char_value='bar')

assert test_obj not in TestObj.objects.filter(Q(m2ms__int_value=10) & Q(m2ms__char_value='bar'))
assert test_obj in TestObj.objects.filter(~(~Q(m2ms__int_value=10) | ~Q(m2ms__char_value='bar')))
```
Pulling down this branch and testing things out in an example application, it seems these counterintuitive assertions still hold. Do you have thoughts on whether obeying De Morgan's law is a desirable invariant for `Q` objects and querysets? I think it would make the ORM querying code easier to reason about from an end-user perspective.

**Comment by akaariai** [comment]
Yes, I believe we should obey De Morgan's laws. Where we can't due to query complexity reasons, we should throw errors instead of producing nonsensible results. The only exception is that we might have managed to publicly document that we do not obey De Morgan for some cases.

The problem here is that we are ending up with extremely complex queries. This PR is draining a lot of time, and it seems it is hard to find time for even small patches these days. But I'll try to get back to this as soon as I can.

**Comment by akaariai** [comment]
Closing due to lack of time to work on this.

**Pull request closed without merging**